oauth/api/router.py

```diff
@@ -0,0 +1,345 @@
+                    self.create_client_handler,
+                    dependencies=[
+                        Security(
+                            self.verify_oauth_client_handler,
+                            scopes=[CLIENT_CREATE],
+                        ),
+                    ],
+                    methods=["POST"],
+                    response_model=ClientCreatedResponse,
+                    status_code=status.HTTP_201_CREATED,
+                    summary="Create a new client",
+                ),
+                APIRoute(
+                    endpoints.CLIENT_BY_ID,
+                    self.delete_client_handler,
+                    dependencies=[
+                        Security(
+                            self.verify_oauth_client_handler,
+                            scopes=[CLIENT_DELETE],
+                        ),
+                    ],
+                    methods=["DELETE"],
+                    response_model=None,  # Explicitly defined, to prevent overwriting
+                    status_code=status.HTTP_204_NO_CONTENT,
+                    summary="Delete a client",
+                ),
+                APIRoute(
+                    endpoints.CLIENT_SCOPE,
+                    self.get_client_scopes_handler,
+                    dependencies=[
+                        Security(
+                            self.verify_oauth_client_handler,
+                            scopes=[CLIENT_READ],
+                        ),
+                    ],
+                    methods=["GET"],
+                    response_model=List[str],
+                    summary="Retrieve the current scopes for a client",
+                ),
+                APIRoute(
+                    endpoints.CLIENT_SCOPE,
+                    self.set_client_scopes_handler,
+                    dependencies=[
+                        Security(
+                            self.verify_oauth_client_handler,
+                            scopes=[CLIENT_UPDATE],
+                        ),
+                    ],
+                    methods=["PUT"],
+                    summary="Overwrite the scopes for an existing client",
+                ),
+                APIRoute(
+                    endpoints.SCOPE,
+                    self.read_scopes,
+                    dependencies=[
+                        Security(
+                            self.verify_oauth_client_handler,
+                            scopes=[SCOPE_READ],
+                        ),
+                    ],
+                    methods=["GET"],
+                    response_model=List[str],
+                    summary="Retrieve all available scopes",
+                ),
+            ],
+        )
+
+    async def acquire_access_token_handler(
+        self,
+        *args: Any,
+        **kwargs: Any,
+    ) -> Callable[..., AccessToken]:
+        """
+        Given a set of credentials, returns an access token.
+        """
+
+        async def acquire_access_token(
```

Comment:
can you tell me a little bit about why you had to wrap a handler around each function?

Reply:
This is a [hacky] workaround to enable the actual handler functions to receive contextual config variables set on the `OAuthRouter` instance. It's not clear to me that it would be possible to pass these arguments if passing the nested functions to the `APIRoute` constructor directly, since the `APIRoute` constructor expects a `Callable`.